New PR and commit guidelines (#2752)
kubevirt/containerized-data-importer: CONTRIBUTING.md

### PR Checklist

Before your PR can be merged it must meet the following criteria:
* Informative PR Description alongside an overview of the changes and a high-level design in case of a new feature
* Commits split logically with a clear purpose.

> * Commits title concise and clear while commit message informative stating the nature and purpose of the changes in the commit.

Reviewer: * Currently our merge policy is squashing all PR commits, so consider:
1. Splitting to incremental small PRs instead of a full feature PR
1. Interactive rebase (git rebase -i) before PR is getting merged, so it won’t look like a list of useless titles

### PR Checklist

Before your PR can be merged it must meet the following criteria:

> * PR Desciption should be informative, explaining what is in it and a high level design in case of new feature

Reviewer 2: Nice! I'm glad we are addressing this issue and I think it's a great idea to update this document. That said, I think we should differentiate between merging criteria (need for documentation, unit tests, etc.) and PR quality guidelines. 

I would add these in a separate section and avoid mixing them with the strict merge criteria. Maybe we could also add examples for each one of the fields (for example, I think https://github.com/kubevirt/containerized-data-importer/pull/2709 would fit nicely as an example for this first point).

Author: I didn't want to overload the lines but an example can be good.
Regarding putting it in a different section Im not sure cause I do think this should be a criteria before merging a PR not just a quality thing.

> * PR Desciption should be informative, explaining what is in it and a high level design in case of new feature

Reviewer 3: ```suggestion
* Informative PR Description alongside an overview of the changes, high-level design in case of new feature
```

If you don't choose to go forward with this, note there's a typo `Desciption`

Author: yeah noted

* Commits should be split logically with a single clear purpose for each.

> * Commits title should be concise and clear while commit message should be as informative as possible.

Author: you are welcome to suggest a new wording for this sentence :)

> * Commits title should be concise and clear while commit message should be as informative as possible.

Reviewer 4: Add a period after informative.  End of sentence.

> * Commits title should be concise and clear while commit message should be as informative as possible.

Reviewer 3: I think we may be able to drop this. I think the idea will get communicated in a softer manner via the previous point

> * Commits title should be concise and clear while commit message should be as informative as possible.

Reviewer 5: Maybe "the commit message should clearly and completely state the nature and purpose of the changes in the commit"

> * Commits title should be concise and clear while commit message should be as informative as possible.

Author: > I think we may be able to drop
> 
> ```
> Commits title should be concise and clear while commit message should be as informative as possible.
> ```
> 
> I think the idea will get communicated in a softer manner via the previous point

I think there is a big difference between how one splits and organizes the commits (the 2nd point) to make sure the title and message of each commit is clear. I have examples in existing PRs for such difference

> * New changes done (after review for example) should be squashed to relevant commit.

Author: It depends cause if a second reviewer is going over the PR after more changes were introduced then he will comment on the commit and then in the newer "updates" commit he might see that everything that he invested time on reviewing was already fix or that his review is already unrelevant

> * New changes done (after review for example) should be squashed to relevant commit.

Reviewer 3: I can see why this is not trivial... harder to review vs harder to make sense of later
but until we start not squashing I lean towards dropping this

> * New changes done (after review for example) should be squashed to relevant commit.

Reviewer 4: > If the majority thinks we should drop it then I will,

@ShellyKa13 the burden is on you to convince the majority.  Not the other way around.  This PR will rot until a significant number of active contributors come forward in support.

> * New changes done (after review for example) should be squashed to relevant commit.

Reviewer 5: I support @ShellyKa13 on this point.  This was a principal part of her arguments.  The commit log of CDI should be clean and changing the same exact lines in multiple commits introduces noise and makes going back to look at individual commits useless.  The thrust of this PR is to "Make `git-log` readable again".

> * New changes done (after review for example) should be squashed to relevant commit.

Reviewer 5: I think you can compare changes between force-pushes using the github UI so even if a PR author squashes their review comment driven changes into the set of existing commits you can still see only what's been changed since the last round of review.
